### PR TITLE
Fix object rest enumeration

### LINKED ISSUE #29616: Object rest includes non-enumerable Symbols
**TypeScript Version:**  3.2.4

**Search Terms:** object rest, rest symbols

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
// Test this by running `tsc` on the command-line, rather than through another build tool such as Gulp, Webpack, etc.

const object = { visible: true };

Object.defineProperty(object, 'does not appear', {
  enumerable: false,
  value: 'hidden key',
});

Object.defineProperty(object, Symbol('appears on rest'), {
  enumerable: false,
  value: 'hidden symbol',
});

// does not include non-enumerable symbol or key
console.log({ ...object });

const { visible, ...rest } = object;

// includes non-enumerable symbol, but excludes key
console.log(rest);
```

**Expected behavior:**

Non-enumerable symbols are not include in Object rest operations.

**Actual behavior:**

Non-enumerable symbols are included in Object rest operations.

**Playground Link:** 

[Playground](https://www.typescriptlang.org/play/#src=%0D%0Aconst%20object%20%3D%20%7B%20visible%3A%20true%20%7D%3B%0D%0A%0D%0AObject.defineProperty(object%2C%20'does%20not%20appear'%2C%20%7B%0D%0A%20%20enumerable%3A%20false%2C%0D%0A%20%20value%3A%20'hidden%20key'%2C%0D%0A%7D)%3B%0D%0A%0D%0AObject.defineProperty(object%2C%20Symbol('appears%20on%20rest')%2C%20%7B%0D%0A%20%20enumerable%3A%20false%2C%0D%0A%20%20value%3A%20'hidden%20symbol'%2C%0D%0A%7D)%3B%0D%0A%0D%0A%2F%2F%20does%20not%20include%20non-enumerable%20symbol%0D%0Aconsole.log(%7B%20...object%20%7D)%3B%0D%0A%0D%0Aconst%20%7B%20visible%2C%20...rest%20%7D%20%3D%20object%3B%0D%0A%0D%0A%2F%2F%20does%20includes%20non-enumerable%20symbol%0D%0Aconsole.log(rest)%3B)

**Related Issues:** 

I believe this was introduced with this PR => https://github.com/Microsoft/TypeScript/pull/12248.

This included Symbols in rest, but uses `Object.getOwnPropertySymbols` without using `Object.prototype.propertyIsEnumerable` check in the loops. You can see this in the generated output in the playground above.

### PR DESCRIPTION
**Pull request by NicholasLYang**
* [x] There is an associated issue that is labeled
  'Bug' or 'help wanted' or is in the Community milestone
* [x] Code is up-to-date with the `master` branch
* [x] You've successfully run `jake runtests` locally
* [x] You've signed the CLA
* [x] There are new or updated unit tests validating the change

Fixes #29616

Adjusts the __rest function in destructuring.ts to check if property is enumerable.

### CODE REVIEW

File: src/compiler/transformers/destructuring.ts [hunk 1]
```diff
@@ -521,8 +521,10 @@ namespace ts {
                 for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                     t[p] = s[p];
                 if (s != null && typeof Object.getOwnPropertySymbols === "function")
-                    for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
-                        t[p[i]] = s[p[i]];
+                    for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
+                        if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable(p[i]))
```
Comment (rbuckton) [inline src/compiler/transformers/destructuring.ts]: This is incorrect as you are passing the wrong receiver. Its highly unlikely `p[i]` is enumerable on `Object.prototype`. You probably mean this:
```suggestion
                        if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
```

**Comment by rbuckton** [comment]
As @ajafff mentioned above, you should also open a PR against https://github.com/Microsoft/tslib for the same change:

- https://github.com/Microsoft/tslib/blob/master/tslib.js#L82
- https://github.com/Microsoft/tslib/blob/master/tslib.es6.js#L46

File: src/compiler/transformers/destructuring.ts (hunk 1 above)
Comment (NicholasLYang) [reply src/compiler/transformers/destructuring.ts]: Ah, makes sense. In that case, does `s.propertyIsEnumerable(p[i])` work as well?

Comment (NicholasLYang) [reply src/compiler/transformers/destructuring.ts]: On second thought, I suppose `s` could have its own implementation of propertyIsEnumerable, so no, that probably does not work.

**Comment by rbuckton** [comment]
It looks like you still have some failing tests.

**Comment by NicholasLYang** [comment]
Yep, sorry. Fixed!

**Pull request merged**